src/main/java/com/uid2/shared/attest/UidCoreClient.java

```diff
@@ -71,179 +59,90 @@ public ICloudStorage getContentStorage() {
         return this.contentStorage;
     }
 
-    public void attest() throws IOException, UidCoreClientException {
-        attestInternal();
+    @Override
+    public InputStream download(String path) throws CloudStorageException {
+        try {
+            InputStream inputStream;
+            if (allowContentFromLocalFileSystem && path.startsWith("file:/tmp/uid2")) {
+                // returns `file:/tmp/uid2` urlConnection directly
+                inputStream = readContentFromLocalFileSystem(path, this.proxy);
+            } else {
+                inputStream = getWithAttest(path);
+            }
+            return inputStream;
+        } catch (Exception e) {
+            throw new CloudStorageException("download " + path + " error: " + e.getMessage(), e);
+        }
     }
 
-    private InputStream getWithAttest(String url) throws IOException, UidCoreClientException {
-        if (!attested())
-            attestInternal();
-
-        URLConnection conn = sendGet(url);
+    private InputStream readContentFromLocalFileSystem(String path, Proxy proxy) throws IOException {
+        return (proxy == null ? new URL(path).openConnection() : new URL(path).openConnection(proxy)).getInputStream();
+    }
 
-        if (conn instanceof HttpURLConnection && attestIfRequired((HttpURLConnection) conn))
-            conn = sendGet(url);
+    private InputStream getWithAttest(String path) throws IOException, InterruptedException, AttestationTokenRetrieverException {
+        if (!attestationTokenRetriever.attested()) {
+            attestationTokenRetriever.attestInternal();
+        }
 
-        return conn.getInputStream();
-    }
+        String attestationToken = attestationTokenRetriever.getAttestationToken();
+        HttpResponse<String> httpResponse;
+        httpResponse = sendHttpRequest(path, attestationToken);
 
-    private boolean attestIfRequired(HttpURLConnection conn) throws IOException, UidCoreClientException {
-        boolean attested = false;
-        int statusCode = conn.getResponseCode();
-        if (statusCode == 401) {
+        if (httpResponse.statusCode() == 401) {
             LOGGER.info("Initial response from UID2 Core returned 401, performing attestation");
-            attested = true;
-            try {
-                attestInternal();
-            }
-            catch (UidCoreClientException | IOException e) {
-                notifyResponseStatusWatcher(statusCode);
-                throw e;
-            }
-        } else {
-            notifyResponseStatusWatcher(statusCode);
+            attestationTokenRetriever.attestInternal();
```

Comment:
This code will probably not get executed unless attestation is failing as you have just called attestInternal above.

Reply:
Talked offline. We agreed on this `if (httpResponse.statusCode() == 401)` clause should never happen, but keeping this part of the code just to be extra safe. Adding a comment to it.